### DISCUSSION TITLE
Issue in relationship

### DISCUSSION THREAD

**Discussion by wasiqaftab** [binaryk/laravel-restify#539] [Q&A] [answered]
HI sorry a wired bug in relationship or may be its me i am scratching my head from last 2 days but didnt get around it. please see the code 
` return [
            'contactsUser' => HasMany::make('userfriends', UserRepository::class),
        ];`

but the  result which i get from index function just one record relationship recorded. can you please tell me what i am doing wrong i see the repo keys etc all are okay. userfriends has a relation in contact model @binaryk can you please check it

**Comment by arthurkirkosa** [comment]
I'd recommend you use the same key in the Repository as the name of the function in your User model.

`return [ 'userfriends' => HasMany::make('userfriends', UserRepository::class)];` if you relation in User is `public function userfriends(): HasMany`

**Comment by wasiqaftab** [comment]
@arthurkirkosa i am doing same please see the snippets

Table structure is 

users have user id as uuid 
contacts table have 
**user_id** as uuid and **friend_id** as uuid 

1. User Model

` public function userfriends(): HasMany
    {
        return $this->hasMany(Contact::class,'friend_id', 'id');
    }`

2. ContactsRepository Relation

Key on contacts repository is `    public static string $uriKey = 'contacts';` 

` public static function related(): array
    {
        return [
            'userfriends' => HasMany::make('userfriends', UserRepository::class),
        ];
}`

3. Contact Model

` public function userfriends(): BelongsTo
    {
        return $this->belongsTo(User::class, 'friend_id');
    }`

API call if is like that 

`{{host}}/api/restify/contacts?related=userfriends`

Result i am expecting is all friend_ids with relation but now i am getting error of un defined function maps 

` "message": "Call to undefined method App\\Models\\User::map()",`

Please guide am i doing something wrong

**Comment by wasiqaftab** [comment]
Also if i use belongsTo relation like that 
` return [
            'userRelationship' => BelongsTo::make('userfriends', UserRepository::class),
        ];`

it will only return first record as relation output something like that

![Screenshot 2023-02-14 at 9 45 19 pm](https://user-images.githubusercontent.com/3602917/218870255-7445fc9d-ec51-4fdf-808f-c1ec8a823f85.png)

Please guide as it just show relation for first record

**Comment by arthurkirkosa** [comment, marked as answer]
Yes, as Contact has a BelongTo with User, that's the relation you have to use in your ContactRepository as well.

Because a Contact Belongs To a User, it's normal for you to have just one item of User loaded with that Contact

If you what the full list of contacts start from the User repo 
`GET /api/restify/users?related=userfriends`

For clarity I would rename the methods in each model to the corresponding entity

```
// User

public function contacts(): HasMany {}
// or
public function friends(): HasMany {}

// Contact

public function user(): BelongsTo {}
// or
public function friend(): BelongsTo {}
```

Naming them in singular and plural would outline how many object I should receive when using the relation in the code

**Comment by arthurkirkosa** [comment]
@wasiqaftab I moved this issue to a Discussion.

**Comment by wasiqaftab** [comment]
@arthurkirkosa i have same issue 
<img width="750" alt="Screenshot 2023-02-15 at 3 34 41 PM" src="https://user-images.githubusercontent.com/3602917/219004427-74c9d004-795a-42a5-995d-5315bea72e9a.png">
<img width="782" alt="Screenshot 2023-02-15 at 3 34 50 PM" src="https://user-images.githubusercontent.com/3602917/219004440-937a48eb-fe34-4efa-aad6-56e027ff5e42.png">
<img width="731" alt="Screenshot 2023-02-15 at 3 35 16 PM" src="https://user-images.githubusercontent.com/3602917/219004447-72730015-eabe-4f5b-9108-6916ef99a336.png">

**Reply by wasiqaftab** [reply]
using that code

**Reply by arthurkirkosa** [reply]
As I can see, you have defined `friend` as a `BelongsTo` in the model and a `HasMany` in the repository... the type must match in both 

`'contactusers' => BelongsTo::make('friend', UserRepository::class),`

**Reply by wasiqaftab** [reply]
Thanks it worked. but problem is that i just need the user which is in my contact tables. like user have many contacts so i just need friend id relation or user detail can you guide me how i can acheive that. because my understanding is that i need to use the contact repository for it so i can only get the loginin user contacts with their details i mean user details

**Reply by wasiqaftab** [reply]
I think so its a other way around. as contacts table have a composite primary key one is the auth key which tells me that get that users and second is its friends key which tells me okay these are the friends of that login user. now i need get the user info for that user friends like their name etc

**Reply by arthurkirkosa** [reply]
Lets back up.

What are you trying to achieve?

What are the HTTP calls you are making?

**Reply by wasiqaftab** [reply]
i am doing this call
`{{host}}/api/restify/users?related=contactusers`

but logically i want to do this call 

`{{host}}/api/restify/contacts?related=contactusers`

What i am trying to do is to get all the contacts with their relation with users like users have contacts and when i login i need to make api for contacts which will give me all friend ids for that user and friend id is logically another user id so i need to get the data for that friend id which is present in user table

**Comment by arthurkirkosa** [comment]
`{{host}}/api/restify/contacts?related=contactusers` should give you only one user back... which is correct, because a Contact has ONLY one User related (it belongs to that User)

When you log in with the current user... it has many contacts, which have a single relation to another User... and you need to see the currently logged in user's friends (Contact->User)

The architecture that you want is a User has many friends (User) and each User can befriend other users.

If so, you need a ManyToMany relationship between users and don't need that Contact model.

**Reply by wasiqaftab** [reply]
let me try

**Reply by wasiqaftab** [reply]
@arthurkirkosa i have done many to many relation all okay now only one issue please see my output and relations

Thats my Db design for contacts table

![Screenshot 2023-02-15 at 8 05 58 pm](https://user-images.githubusercontent.com/3602917/219141018-9774ec18-7139-445b-a68c-a18c02f65e47.png)

user model 
`public function contacts() : BelongsToMany
    {
        return $this->belongsToMany(Contact::class, 'contacts', 'user_id','friend_id');
    }`

contacts model

`  public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'contacts', 'user_id', 'friend_id');
    }`

contacts repository

` public static function related(): array
    {
        return [
           'contacts' => BelongsToMany::make('users', UserRepository::class),
        ];
    }`

logically contacts is a pivot but user is the main table sorry my bad as i never occur that kind of Sanrio if you see the out not it shows one friend and against that friend all users

![Screenshot 2023-02-15 at 8 13 15 pm](https://user-images.githubusercontent.com/3602917/219142412-adfdbb39-6e85-4aae-88bb-cc90166ec8c8.png)

if you see here it fetch all the users friends data which is write but show the other records same records i mean without relation as well. can you please guide i am willing to pay a bounty on this one as well. because i have 3 to 4 same Sanrio like this in my application.

also for logined use i added indexquery with auth user something like that so i only get records for the user who is login. 

`$query->where('user_id','=',auth()->user()->id)->get();`

**Reply by wasiqaftab** [reply]
I think so i figure it out. now it works my bad i am using the contact repository i need to use the user repository

**Reply by wasiqaftab** [reply]
Only last thing i have about 3 different scenarios which uses the different Pivot tables like block users and profile views. so i am declaring it like that in user repository 
` return [
            'users' => BelongsToMany::make('liked', UserRepository::class)->withPivot(
                field('is_mutual_like')
            ),
            'usersviews' => BelongsToMany::make('viewed', UserRepository::class)->withPivot(
                field('count')
            ),
        ];`

but when i fetch it like 

`{{host}}/api/restify/users/dfb6ec7e-e37b-4cd1-bdf9-ffd3727820f8?related=usersviews`

it gives me both relations if i need to get only one relation at a time how can i do that. and also how i can i include paging in it.

**Reply by wasiqaftab** [reply]
also i try to do the pagination like 
`{{host}}/api/restify/users/dfb6ec7e-e37b-4cd1-bdf9-ffd3727820f8?related=usersviews&relatablePerPage=2`
it will only show 2 records ? as i have 10 records in db

**Reply by arthurkirkosa** [reply]
First, I'd name both the key of the related and the name of the relationship the same; this way you're sure to not load things by mistake or confuse Restify.
```
'liked' => BelongsToMany::make('liked', UserRepository::class)->withPivot( field('is_mutual_like'),
```
if `liked` is the name of the BelongsToMany relationship in the model

**Reply by wasiqaftab** [reply]
Ok thanks it worked. What about paging issue
